Rename arrow module to stochastic_arrow

thalassemia opened the pull request:
Fixes #51.

1fish2 commented on README.md:
```diff
@@ -95,26 +94,30 @@ history = reenact_events(stoichiometry, result['events'], state)
 
+### Version 0.5.3
+
+* Rename module to `stochastic_arrow` to avoid name conflict (Issue #51).
```
* A totally breaking change to clients calls for a major version number change. So let's go to `1.0.0`.
* Let's make this change note bolder: Every project needs to update their import statements, changing `arrow` to `stochastic_arrow`.

* Let's remember to test that `pip install -U` from arrow 0.5.2 to this new release (1) successfully removes the `arrow/` directory from site-packages in the case where the `arrow` date-time package was not installed, and (2) doesn't remove it in the case where the `arrow` date-time package installed its code into the `arrow/` directory. -- If this test fails, maybe there's a workaround in `setup.py`. Otherwise we'll have to give manual workaround instructions to people.

thalassemia replied:
Good idea. While upgrading directly to this release does handle (1), it fails (2). I added a note to the README telling people to uninstall `arrow` before upgrading this package, then reinstall it. I tried automating this in `setup.py` by using `subprocess` to call `pip` but could not get it to work when installing from a source distribution. Plus it seemed like a very bad idea anyways.

1fish2 reviewed (commented):
This looks great! A few deets, below.

Additional files to update:
* Makefile -- needs to clean built files from the new directory. I don't have an opinion on whether it should clean the obsolete arrow/ directory or leave that to developers. I used git clean, which is a bit risky, e.g. it deleted the .python-version file.
* MANIFEST.in
* stochastic_arrow/arrow.pxd -- a comment
* stochastic_arrow/__init__.py
* stochastic_arrow/obsidian.c -- printf calls

thalassemia commented:
I added `pytest.raises` to the flagella test as you suggested. I also wrote a new test to check that the reaction propensities are printed in accordance with the flagella test commit (c91d6fe). Lastly, I added a Github workflow file to automate testing.

thalassemia commented:
To get the Windows tests passing, I had to:

1. Explicitly tell `setup.py` to use `utf-8` to decode the README
2. Explicitly type Numpy arrays as `np.int64`
3. Change the Cython imports from `libc.stdint` to `numpy` types
4. Raise the number of allowed memory increases to 10

Funnily enough, all the Windows test logs have `-nan(ind)`, which is even crazier than the Linux output :satisfied:

1fish2 commented:
That's a lot of wrangling. Can GitHub actions run macOS? Anyway, `make clean compile` now prints lots of warnings like this:

    stochastic_arrow/arrowhead.c:3128:36: warning: incompatible pointer types assigning to 'int64_t *' (aka 'long long *') from '__pyx_t_5numpy_int64_t *' (aka 'long *') [-Wincompatible-pointer-types]

That line of C comes from stochastic_arrow/arrowhead.pyx:75

    self.info.reactions = &reactions[0]

Perhaps all the C code needs to include numpy's types like `int64` rather than `<stdint.h>`? It looks like that's in `mersenne.h` and `arrowhead.c`.

thalassemia commented:
Good call. Number 3 on my list of changes above was actually unnecessary. Removing it gets rid of the incompatible pointer types warnings without breaking any of the Windows tests. I still see errors like this on the macOS runners, but they seem to be a [known Cython issue](https://github.com/cython/cython/issues/5041) with typed memory views:

```
stochastic_arrow/arrowhead.c:22051:21: warning: fallthrough annotation in unreachable code [-Wunreachable-code-fallthrough]
                    CYTHON_FALLTHROUGH;
                    ^
```

1fish2 reviewed (approved):
LGTM!

Yes, we've been ignoring CYTHON_FALLTHROUGH warnings for a while. Eventually Cython will fix that.

1fish2 commented:
@thalassemia Do you want to tune anything else, or merge this PR, or have me merge it?

thalassemia commented:
There were a few type-casting warnings in the Windows test logs that bothered me. I spent the morning getting my Windows environment set up and think they should be ironed out now👍

thalassemia commented:
All the warnings have been resolved. I'll go ahead and merge now.
